File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/NestedBorrowedFields.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;
import java.lang.foreign.SegmentAllocator;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class NestedBorrowedFields {
    BorrowedFields fields;
    BorrowedFieldsWithBounds bounds;
    BorrowedFieldsWithBounds bounds2;
    

    List<Object> selfEdges = List.of();
    List<Object> xEdges = List.of();
    List<Object> yEdges = List.of();
    List<Object> zEdges = List.of();
    

    private NestedBorrowedFields() {
    }

    NestedBorrowedFields(MemorySegment structSegment, List<Object> xEdges, List<Object> yEdges, List<Object> zEdges) {
        this.selfEdges = selfEdges;
        this.xEdges = xEdges;
        this.yEdges = yEdges;
        this.zEdges = zEdges;
        

        var fieldsNative = dev.diplomattest.somelib.ntv.NestedBorrowedFields.fields(structSegment);
        var fieldsVal = new BorrowedFields(fieldsNative, List.of(xEdges, yEdges));
        this.fields = fieldsVal;
        var boundsNative = dev.diplomattest.somelib.ntv.NestedBorrowedFields.bounds(structSegment);
        var boundsVal = new BorrowedFieldsWithBounds(boundsNative, List.of(xEdges, yEdges), List.of(), List.of());
        this.bounds = boundsVal;
        var bounds2Native = dev.diplomattest.somelib.ntv.NestedBorrowedFields.bounds2(structSegment);
        var bounds2Val = new BorrowedFieldsWithBounds(bounds2Native, List.of(), List.of(), List.of());
        this.bounds2 = bounds2Val;
        

    }

    MemorySegment toNative(SegmentAllocator arena) {
        var returnVal = dev.diplomattest.somelib.ntv.NestedBorrowedFields.allocate(arena);
        
        var fieldsNative = fields.toNative(arena);
        dev.diplomattest.somelib.ntv.NestedBorrowedFields.fields(returnVal, fieldsNative);
        var boundsNative = bounds.toNative(arena);
        dev.diplomattest.somelib.ntv.NestedBorrowedFields.bounds(returnVal, boundsNative);
        var bounds2Native = bounds2.toNative(arena);
        dev.diplomattest.somelib.ntv.NestedBorrowedFields.bounds2(returnVal, bounds2Native);
        

        return returnVal;

    }
    
    public static NestedBorrowedFields fromBarAndFooAndStrings(Bar bar,Foo foo,String dstr16X,String dstr16Z,String utf8StrY,String utf8StrZ) {
        
        try (var arena = Arena.ofConfined()) {
            
            var barNative = bar.internal;
            var fooNative = foo.internal;
            var dstr16XData = Arena.ofAuto().allocateFrom(dstr16X, StandardCharsets.UTF_16);
            var dstr16XLen = dstr16XData.byteSize() - 1;  // allocated strings are null terminated
            var dstr16XView = DiplomatString16View.allocate(Arena.ofAuto());
            DiplomatString16View.len(dstr16XView, dstr16XLen);
            DiplomatString16View.data(dstr16XView, dstr16XData);
            var dstr16ZData = Arena.ofAuto().allocateFrom(dstr16Z, StandardCharsets.UTF_16);
            var dstr16ZLen = dstr16ZData.byteSize() - 1;  // allocated strings are null terminated
            var dstr16ZView = DiplomatString16View.allocate(Arena.ofAuto());
            DiplomatString16View.len(dstr16ZView, dstr16ZLen);
            DiplomatString16View.data(dstr16ZView, dstr16ZData);
            var utf8StrYData= Arena.ofAuto().allocateFrom(utf8StrY, StandardCharsets.UTF_8);
            var utf8StrYLen = utf8StrYData.byteSize() - 1;  // allocated strings are null terminated
            var utf8StrYView = DiplomatStringView.allocate(Arena.ofAuto());
            DiplomatStringView.len(utf8StrYView, utf8StrYLen);
            DiplomatStringView.data(utf8StrYView, utf8StrYData);
            var utf8StrZData= Arena.ofAuto().allocateFrom(utf8StrZ, StandardCharsets.UTF_8);
            var utf8StrZLen = utf8StrZData.byteSize() - 1;  // allocated strings are null terminated
            var utf8StrZView = DiplomatStringView.allocate(Arena.ofAuto());
            DiplomatStringView.len(utf8StrZView, utf8StrZLen);
            DiplomatStringView.data(utf8StrZView, utf8StrZData);
            var nativeVal = somelib_h.NestedBorrowedFields_from_bar_and_foo_and_strings(arena, barNative, fooNative, dstr16XView, dstr16ZView, utf8StrYView, utf8StrZView);
            
            var returnVal = new NestedBorrowedFields(nativeVal, List.of(bar, dstr16X, utf8StrY), List.of(bar, utf8StrY), List.of(foo, dstr16Z, utf8StrZ));
            return returnVal;
                    
        }
    }
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/NestedBorrowedFields.kt
package dev.diplomattest.somelib

import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface NestedBorrowedFieldsLib: Library {
    fun NestedBorrowedFields_from_bar_and_foo_and_strings(bar: Pointer, foo: Pointer, dstr16X: Slice, dstr16Z: Slice, utf8StrY: Slice, utf8StrZ: Slice): NestedBorrowedFieldsNative
}

internal class NestedBorrowedFieldsNative: Structure(), Structure.ByValue {
    @JvmField
    internal var fields: BorrowedFieldsNative = BorrowedFieldsNative();
    @JvmField
    internal var bounds: BorrowedFieldsWithBoundsNative = BorrowedFieldsWithBoundsNative();
    @JvmField
    internal var bounds2: BorrowedFieldsWithBoundsNative = BorrowedFieldsWithBoundsNative();

    // Define the fields of the struct
    override fun getFieldOrder(): List<String> {
        return listOf("fields", "bounds", "bounds2")
    }
}




internal class OptionNestedBorrowedFieldsNative constructor(): Structure(), Structure.ByValue {
    @JvmField
    internal var value: NestedBorrowedFieldsNative = NestedBorrowedFieldsNative()

    @JvmField
    internal var isOk: Byte = 0

    // Define the fields of the struct
    override fun getFieldOrder(): List<String> {
        return listOf("value", "isOk")
    }

    internal fun option(): NestedBorrowedFieldsNative? {
        if (isOk == 1.toByte()) {
            return value
        } else {
            return null
        }
    }


    constructor(value: NestedBorrowedFieldsNative, isOk: Byte): this() {
        this.value = value
        this.isOk = isOk
    }

    companion object {
        internal fun some(value: NestedBorrowedFieldsNative): OptionNestedBorrowedFieldsNative {
            return OptionNestedBorrowedFieldsNative(value, 1)
        }

        internal fun none(): OptionNestedBorrowedFieldsNative {
            return OptionNestedBorrowedFieldsNative(NestedBorrowedFieldsNative(), 0)
        }
    }

}

class NestedBorrowedFields (var fields: BorrowedFields, var bounds: BorrowedFieldsWithBounds, var bounds2: BorrowedFieldsWithBounds) {
    companion object {

        internal val libClass: Class<NestedBorrowedFieldsLib> = NestedBorrowedFieldsLib::class.java
        internal val lib: NestedBorrowedFieldsLib = Native.load("diplomat_feature_tests", libClass)
        val NATIVESIZE: Long = Native.getNativeSize(NestedBorrowedFieldsNative::class.java).toLong()

        internal fun fromNative(nativeStruct: NestedBorrowedFieldsNative, xEdges: List<Any?>, yEdges: List<Any?>, zEdges: List<Any?>): NestedBorrowedFields {
            val fields: BorrowedFields = BorrowedFields.fromNative(nativeStruct.fields, xEdges)
            val bounds: BorrowedFieldsWithBounds = BorrowedFieldsWithBounds.fromNative(nativeStruct.bounds, xEdges, yEdges, yEdges)
            val bounds2: BorrowedFieldsWithBounds = BorrowedFieldsWithBounds.fromNative(nativeStruct.bounds2, zEdges, zEdges, zEdges)

            return NestedBorrowedFields(fields, bounds, bounds2)
        }

        @JvmStatic
        
        fun fromBarAndFooAndStrings(bar: Bar, foo: Foo, dstr16X: String, dstr16Z: String, utf8StrY: String, utf8StrZ: String): NestedBorrowedFields {
            // This lifetime edge depends on lifetimes: 'x, 'y
            val xEdges: MutableList<Any> = mutableListOf(bar);
            // This lifetime edge depends on lifetimes: 'y
            val yEdges: MutableList<Any> = mutableListOf(bar);
            // This lifetime edge depends on lifetimes: 'z
            val zEdges: MutableList<Any> = mutableListOf(foo);
            val dstr16XSliceMemory = PrimitiveArrayTools.borrowUtf16(dstr16X).into(listOf(xEdges))
            val dstr16ZSliceMemory = PrimitiveArrayTools.borrowUtf16(dstr16Z).into(listOf(zEdges))
            val utf8StrYSliceMemory = PrimitiveArrayTools.borrowUtf8(utf8StrY).into(listOf(yEdges))
            val utf8StrZSliceMemory = PrimitiveArrayTools.borrowUtf8(utf8StrZ).into(listOf(zEdges))
            
            val returnVal = lib.NestedBorrowedFields_from_bar_and_foo_and_strings(bar.handle, foo.handle, dstr16XSliceMemory.slice, dstr16ZSliceMemory.slice, utf8StrYSliceMemory.slice, utf8StrZSliceMemory.slice);
            val returnStruct = NestedBorrowedFields.fromNative(returnVal, xEdges, yEdges, zEdges)
            return returnStruct
        }
    }
    internal fun toNative(xAppendArray: Array<MutableList<Any>>, yAppendArray: Array<MutableList<Any>>, zAppendArray: Array<MutableList<Any>>): NestedBorrowedFieldsNative {
        var native = NestedBorrowedFieldsNative()
        native.fields = this.fields.toNative(aAppendArray = xAppendArray)
        native.bounds = this.bounds.toNative(aAppendArray = xAppendArray, bAppendArray = yAppendArray, cAppendArray = yAppendArray)
        native.bounds2 = this.bounds2.toNative(aAppendArray = zAppendArray, bAppendArray = zAppendArray, cAppendArray = zAppendArray)
        return native
    }

    internal fun xEdges(): List<Any?> {
        return TODO("todo")
    }
    internal fun yEdges(): List<Any?> {
        return TODO("todo")
    }
    internal fun zEdges(): List<Any?> {
        return TODO("todo")
    }
}